### PR TITLE
fix mul and imul for *dx operands and 64 bit widths ##esil

### PR DESCRIPTION
**Pull request by aemmitt-ns**
**Checklist**

- [x] Closing issues: #18860
- [ ] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [radare2book](https://github.com/radareorg/radare2book)

**Description**

Fixes my bad mul / imul ESIL which did not work for *dx operands (the register is written to and then used again as if it were not) and also MUL was not valid for 64 bits. 64 bit IMUL instructions will still be wrong since theres no 128 bit sign extension. I will fix this at some point. I really hate IMUL.

### CODE REVIEW

**Comment by aemmitt-ns** [comment]
I am going to fix the tests and also refactor the IMUL code. its gotten real ugly, mostly because of me

**Comment by trufae** [comment]
Can you update the tests and add a couple of tests emulating and checking the registers afterwards? Having yests with the esil string around is not much useful i think.

we all hate mul and div :{{ totally not pleasant to describe

**Comment by aemmitt-ns** [comment]
ok still going to add some emulation tests. i agree that the string tests aren't super useful. I'd be ok with getting rid of them in general and relying only on tests of the emulation results.

**Comment by aemmitt-ns** [comment]
Alright I think this is ready to go now. Added tests for the relevant instructions and everything is passing now

File: libr/anal/p/anal_x86_cs.c [hunk 1]
```diff
@@ -1697,29 +1697,26 @@ static void anop_esil(RAnal *a, RAnalOp *op, ut64 addr, const ut8 *buf, int len,
...
-						}
-					}
-					else {
-						/* should never happen */
+				if (arg0) {
+					const char *r_quot = (width==1)?"al": (width==2)?"ax": (width==4)?"eax":"rax";
+					const char *r_rema = (width==1)?"ah": (width==2)?"dx": (width==4)?"edx":"rdx";
+					const char *r_nume = (width==1)?"ax": r_quot;
+
+					if ( width == 8 ) { // TODO still needs to be fixed to handle correct signed 128 bit value
```
Comment (trufae) [inline libr/anal/p/anal_x86_cs.c:1714]:
```suggestion
					if (width == 8) { // TODO still needs to be fixed to handle correct signed 128 bit value 
```

**Review by trufae — APPROVED** [review]

**Pull request merged**